### PR TITLE
Added openstack instance metadata search order

### LINKED ISSUE #52378: Openstack config drive as source of truth for instance metadata
**Is this a BUG REPORT or FEATURE REQUEST?**: FEATURE REQUEST
/kind feature

**What happened**:
Openstack's config drive is used as source of truth for instance metadata. Config drive gets out of date pretty quick (as soon as the instance name is updated for example), this will result in an error when trying to start the kubelet service cause the name on the metadata (outdated now) won't match the openstack instance list.

**What you expected to happen**:
Allow the user to disable the use of openstack's config drive as sole source of truth. In other words force a search on the metadata service. 

**How to reproduce it (as minimally and precisely as possible)**:

1. Configure kubelet service to use openstack as the cloud provider
3. `nova boot --config-drive true --flavor ... --image ... --security-groups "..." --key-name "..." --nic net-id=... currentname`
2. `nova update --name newname currentname`
3. Restart kubelet service

**Anything else we need to know?**:
The restriction is on the order of these calls: https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/openstack/metadata.go#L146

**Environment**:
- Kubernetes version (use `kubectl version`):
`Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.4", GitCommit:"793658f2d7ca7f064d2bdf606519f9fe1229c381", GitTreeState:"clean", BuildDate:"2017-08-17T17:03:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.1", GitCommit:"1dc5c66f5dd61da08412a74221ecc79208c2165b", GitTreeState:"clean", BuildDate:"2017-07-14T01:48:01Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}`
- Cloud provider or hardware configuration**:
Openstack
- OS (e.g. from /etc/os-release):
`NAME="Ubuntu"
VERSION="14.04.5 LTS, Trusty Tahr"`
- Kernel (e.g. `uname -a`):
`Linux 1b24f0b2-c15a-4c67-aef2-be6a0fe378fe 4.4.0-83-generic #106~14.04.1-Ubuntu SMP Mon Jun 26 18:10:19 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`

### PR DESCRIPTION
**Pull request by richardalberto**
**What this PR does / why we need it**: This PR adds a search order for the instance metadata retrieval on openstack. More information and discussion can be found on #52378 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #52378 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by richardalberto** [comment]
signed CLA

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 1]
```diff
@@ -742,3 +755,26 @@ func (os *OpenStack) volumeService(forceVersion string) (volumeService, error) {
 		return nil, errors.New(err_txt)
 	}
 }
+
+func checkMetadataSearchOrder(order []string) error {
+	if order == nil {
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: We should default to what we do currently for backward compat purposes.

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 2]
```diff
@@ -32,17 +32,27 @@ import (
...
-const configDrivePath = "openstack/2012-08-10/meta_data.json"
+const (
+	// metadataUrl is URL to OpenStack metadata server. It's hardcoded IPv4
+	// link-local address as documented in "OpenStack Cloud Administrator Guide",
+	// chapter Compute - Networking with nova-network.
+	// https://docs.openstack.org/admin-guide/compute-networking-nova.html#metadata-service
+	metadataUrl = "http://169.254.169.254/openstack/2012-08-10/meta_data.json"
+
+	// metadataID is used as an identifier on the metadata search order configuration.
+	metdataID = "metadataservice"
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/metadata.go]: typo `metadataID`

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 1 above)
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/openstack.go]: I'm currently setting the default on line 287. This conditional is for an scenario that can't happen on the current code, but keeps the function consistent by not trusting it's input.

File: pkg/cloudprovider/providers/openstack/metadata.go (hunk 2 above)
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/metadata.go]: Good catch! corrected.

**Comment by richardalberto** [comment]
is there anything else I can do to get this PR merged?

**Comment by dims** [comment]
cc @FengyunPan

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 1 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack.go]: To read easily, can you set the default calue in readConfig()?

Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack.go]: To read easily, can you set the default value in readConfig()?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 3]
```diff
@@ -275,16 +281,23 @@ func newOpenStack(cfg Config) (*OpenStack, error) {
 		err = openstack.Authenticate(provider, cfg.toAuthOptions())
 	}
 
+	if cfg.Metadata.SearchOrder == nil {
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack.go]: '== nil' ------> '!= nil'

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 4]
```diff
@@ -32,17 +32,27 @@ import (
...
-const configDrivePath = "openstack/2012-08-10/meta_data.json"
+const (
+	// metadataUrl is URL to OpenStack metadata server. It's hardcoded IPv4
+	// link-local address as documented in "OpenStack Cloud Administrator Guide",
+	// chapter Compute - Networking with nova-network.
+	// https://docs.openstack.org/admin-guide/compute-networking-nova.html#metadata-service
+	metadataUrl = "http://169.254.169.254/openstack/2012-08-10/meta_data.json"
+
+	// metadataID is used as an identifier on the metadata search order configuration.
+	metadataID = "metadataservice"
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/metadata.go]: 'metadataservice' -->'metadataService'

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 5]
```diff
@@ -32,17 +32,27 @@ import (
...
+
+const (
+	// Config drive is defined as an iso9660 or vfat (deprecated) drive
+	// with the "config-2" label.
+	// http://docs.openstack.org/user-guide/cli-config-drive.html
+	configDriveLabel = "config-2"
+	configDrivePath  = "openstack/2012-08-10/meta_data.json"
+
+	// configDriveID is used as an identifier on the metadata search order configuration.
+	configDriveID = "configdrive"
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/metadata.go]: 'configdrive'--> 'configDrive'

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 6]
```diff
@@ -141,12 +151,24 @@ func getMetadataFromMetadataService() (*Metadata, error) {
...
-			md, err = getMetadataFromMetadataService()
+		var md *Metadata
+		var err error
+
+		for _, id := range order {
+			switch id {
+			case configDriveID:
+				md, err = getMetadataFromConfigDrive()
+			case metadataID:
+				md, err = getMetadataFromMetadataService()
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/metadata.go]: Can we add 'default' case? If the order isn't configDrive and metadataService, err will be nil and break, that's bad.

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 7]
```diff
@@ -100,13 +100,18 @@ type RouterOpts struct {
 	RouterId string `gcfg:"router-id"` // required
 }
 
+type MetadataOpts struct {
+	SearchOrder []string `gcfg:"search-order"`
```
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack.go]: Is this('[]string') easy to use for user?

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 3 above)
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/openstack.go]: Good catch! Corrected.

File: pkg/cloudprovider/providers/openstack/metadata.go (hunk 4 above)
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/metadata.go]: Renamed

File: pkg/cloudprovider/providers/openstack/metadata.go (hunk 5 above)
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/metadata.go]: Renamed

File: pkg/cloudprovider/providers/openstack/metadata.go (hunk 6 above)
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/metadata.go]: Added a default option.

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 7 above)
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/openstack.go]: `[]string` would be the standard for ini config types. But we can also use a comma separated value. Which do you think it's more user friendly?

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 1 above)
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/openstack.go]: moved the default value to `readConfig()`

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 7 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack.go]: It is better to use a comma separated.

**Comment by richardalberto** [comment]
@FengyunPan switched to a comma separated configuration.

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 8]
```diff
@@ -141,12 +151,27 @@ func getMetadataFromMetadataService() (*Metadata, error) {
...
+func getMetadata(order string) (*Metadata, error) {
 	if metadataCache == nil {
-		md, err := getMetadataFromConfigDrive()
-		if err != nil {
-			md, err = getMetadataFromMetadataService()
+		var md *Metadata
+		var err error
+
+		elements := strings.Split(order, ",")
+		for _, id := range elements {
```
Comment (NickrenREN) [inline pkg/cloudprovider/providers/openstack/metadata.go]: Any difference with the previous code?  if `getMetadataFromConfigDrive` returns no error, `break` will be executed

Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/metadata.go]: and that's fine, but now we can select the order in which `getMetadataFromConfigDrive` and `getMetadataFromMetadataService` are executed. In an env where we can't trust `getMetadataFromConfigDrive` or should be used as a secondary search option we can define so on `cloud.conf`.

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 9]
```diff
@@ -32,17 +32,27 @@ import (
...
+	// link-local address as documented in "OpenStack Cloud Administrator Guide",
+	// chapter Compute - Networking with nova-network.
+	// https://docs.openstack.org/admin-guide/compute-networking-nova.html#metadata-service
+	metadataUrl = "http://169.254.169.254/openstack/2012-08-10/meta_data.json"
+
+	// metadataID is used as an identifier on the metadata search order configuration.
+	metadataID = "metadataService"
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (richardalberto) [reply pkg/cloudprovider/providers/openstack/metadata.go]: No reason, I agree we can put them together. Will make the change.

File: pkg/cloudprovider/providers/openstack/metadata.go (hunk 8 above)
Comment (NickrenREN) [reply pkg/cloudprovider/providers/openstack/metadata.go]: Ok. i see. So you mean adding another param to the openstack config file

**Review by NickrenREN — APPROVED** [review]

**Comment by NickrenREN** [comment]
Could you also add some UTs , and then i will add lgtm

**Comment by NickrenREN** [comment]
Actually we need to try to not add param to config file. It will become more and more complex for users to config. 
@richardalberto  any ideas on that ?
cc @anguslees

**Comment by richardalberto** [comment]
@NickrenREN is there any other place where we can configure the cloud provider?

**Comment by richardalberto** [comment]
I think it's also important to consider this is an optional flag.

**Comment by NickrenREN** [comment]
@richardalberto  i do not know where else to configure this since cloud providers do not have their own flags.  Anyway we can go on for this 

/approve

**Comment by NickrenREN** [comment]
Any test cases for this and then lgtm

**Pull request merged**